contracts/XVSVault/XVSVault.sol

```diff
@@ -0,0 +1,376 @@
+     */
+    function deposit(address _rewardToken, uint256 _pid, uint256 _amount) public nonReentrant {
+        PoolInfo storage pool = poolInfos[_rewardToken][_pid];
+        UserInfo storage user = userInfos[_rewardToken][_pid][msg.sender];
+        updatePool(_rewardToken, _pid);
+        if (user.amount > 0) {
+            uint256 pending =
+                user.amount.mul(pool.accRewardPerShare).div(1e12).sub(
+                    user.rewardDebt
+                );
+            IXVSStore(xvsStore).safeRewardTransfer(_rewardToken, msg.sender, pending);
+        }
+        pool.token.transferFrom(
+            address(msg.sender),
+            address(this),
+            _amount
+        );
+        user.amount = user.amount.add(_amount);
+        user.rewardDebt = user.amount.mul(pool.accRewardPerShare).div(1e12);
+        emit Deposit(msg.sender, _rewardToken, _pid, _amount);
+    }
+
+    /**
+     * @notice Execute withdrawal to XVSVault for XVS allocation
+     * @param _rewardToken The Reward Token Address
+     * @param _pid The Pool Index
+     */
+    function executeWithdrawal(address _rewardToken, uint256 _pid) public nonReentrant {
+        PoolInfo storage pool = poolInfos[_rewardToken][_pid];
+        UserInfo storage user = userInfos[_rewardToken][_pid][msg.sender];
+        WithdrawalInfo storage withdrawal = withdrawalInfos[_rewardToken][_pid][msg.sender];
+        uint256 curTimestamp = block.timestamp;
+        uint256 _amount = withdrawal.amount;
+
+        require(withdrawal.amount > 0, "no request to execute");
+        require(lockPeriod.add(withdrawal.timestamp) < curTimestamp, "your request is locked yet");
+
+        updatePool(_rewardToken, _pid);
+        uint256 pending =
+            user.amount.mul(pool.accRewardPerShare).div(1e12).sub(
+                user.rewardDebt
+            );
+        IXVSStore(xvsStore).safeRewardTransfer(_rewardToken, msg.sender, pending);
+        user.amount = user.amount.sub(_amount);
+        user.rewardDebt = user.amount.mul(pool.accRewardPerShare).div(1e12);
+        pool.token.transfer(address(msg.sender), _amount);
+
+        withdrawal.amount = 0;
+
+        emit ExecutedWithdrawal(msg.sender, _rewardToken, _pid, _amount);
+    }
+
+    /**
+     * @notice Request withdrawal to XVSVault for XVS allocation
+     * @param _rewardToken The Reward Token Address
+     * @param _pid The Pool Index
+     * @param _amount The amount to withdraw to vault
+     */
+    function requestWithdrawal(address _rewardToken, uint256 _pid, uint256 _amount) public nonReentrant {
+        UserInfo storage user = userInfos[_rewardToken][_pid][msg.sender];
+        WithdrawalInfo storage withdrawal = withdrawalInfos[_rewardToken][_pid][msg.sender];
+        require(_amount > 0, "requested amount cant be zero");
+        require(user.amount >= _amount, "requested amount is invalid");
+        require(withdrawal.amount == 0, "request again after execute");
```

Comment:
users can make several withdrawal request. we need to update some workflows and storages to implement this. @ferrari105